client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,103 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+function fetchRecent() {
+  return Query.recent().$promise
+    .then((results) => {
+      const filteredResults = results.filter(item => !item.is_draft); // TODO: filter at the fetch level
+      return Promise.resolve(filteredResults);
+    });
+}
+
+function search(q) {
+  return new Promise((resolve) => {
+    Query.query({ q }, ({ results }) => {
+      resolve(results);
+    });
+  });
+}
+
+export default function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  // on search term changed, debounced
+  useEffect(() => {
+    setSearching(true);
+    const debounce = setTimeout(() => {
+      const promise = searchTerm ? () => search(searchTerm) : fetchRecent;
+      promise()
+        .then(setSearchResults)
```

Comment:
It's possible you will receive older search results before newer ones, so you need to check if this is the result of the current `searchTerm`.

Reply:
Right, I put that part aside but forgot to re-introduce it.
I'd like to do this in a more standard way - `response.term` that would come from the API response. Wdyt?

Reply:
Another way to go about it is to disallow concurrent fetches.

Reply:
Or promise.cancel of course.

Reply:
promise.cancel: see the discussion I have with @kravets-levko about this on Slack.

`response.term`: 👍